File: array-string/roman-to-integer/roman-to-integer.ts
function romanToInt(s: string): number {
    const val = (c: string): number => {
        switch (c) {
            case 'I': return 1;
            case 'V': return 5;
            case 'X': return 10;
            case 'L': return 50;
            case 'C': return 100;
            case 'D': return 500;
            case 'M': return 1000;
            default: throw new Error(`Unexpected roman char: '${c}'`);
        }
    };

    let total = 0;
    let prev = 0;

    for (let i = s.length - 1; i >= 0; i--) {
        const v = val(s[i]);
        if (v < prev)
            total -= v;
        else
            total += v;
        prev = v;
    }
    
    return total;
};

File: array-string/roman-to-integer/roman-to-integer.go
func romanToInt(s string) int {
    val := func(c byte) (int, error) {
        switch c {
		case 'I':
			return 1, nil
		case 'V': 
			return 5, nil
		case 'X':
			return 10, nil
		case 'L':
			return 50, nil
		case 'C':
			return 100, nil
		case 'D':
			return 500, nil
		case 'M':
			return 1000, nil
		default: 
			return 0, errors.New(fmt.Sprintf(`Unexpected roman char: %c`, c))
        }
    }

    result := 0;
    prev := 0;

    for i := len(s) - 1; i >= 0; i-- {
        v, _ := val(s[i]);
        if v < prev {
            result -= v;
        } else {
            result += v;
        }
        prev = v;
    }
    
    return result;
}